modules/auxiliary/gather/ldap_esc_vulnerable_cert_finder.rb

```diff
@@ -315,6 +353,107 @@ def find_esc3_vuln_cert_templates
     query_ldap_server_certificates(esc3_template_2_raw_filter, 'ESC3_TEMPLATE_2')
   end
 
+  def find_esc4_vuln_cert_templates
+    # Determine who we are authenticating with. Retrieve the username and user SID
+    whoami_response = @ldap.ldapwhoami
+    username = whoami_response.split('\\')[1]
+
+    user_raw_filter = "(sAMAccountName=#{username})"
+    attributes = ['objectSID']
+    response_user_sid = query_ldap_server(user_raw_filter, attributes)
+    if response_user_sid.empty?
+      print_error("Unable to determine the User SID for #{username}, ESC4 detection can't continue")
+      return
+    else
+      user_sid = Rex::Proto::MsDtyp::MsDtypSid.read(response_user_sid[0][:objectsid].first).value
+    end
+
+    # Perform an LDAP query to get the groups the user is a part of
+    # Use LDAP_MATCHING_RULE_IN_CHAIN OID in order to walk the chain of ancestry of groups.
+    # https://learn.microsoft.com/en-us/windows/win32/adsi/search-filter-syntax?redirectedfrom=MSDN
+    filter_with_user = <<~FILTER
+      (&(member:1.2.840.113556.1.4.1941:=CN=#{username},cn=Users,#{@base_dn}))
+    FILTER
+
+    attributes = ['cn', 'objectSID']
+    esc_entries = query_ldap_server(filter_with_user, attributes)
+
+    user_groups = []
+    esc_entries.each do |entry|
+      print_status(entry[:cn].to_s)
+      group_sid = Rex::Proto::MsDtyp::MsDtypSid.read(entry['ObjectSid'].first).value
+      user_groups << group_sid
+    end
+
+    user_groups << Rex::Proto::Secauthz::WellKnownSids::SECURITY_AUTHENTICATED_USER_SID
+
+    # Determine what Certificate Templates are available to us
+    esc_raw_filter = <<~FILTER
+      (&
+        (objectclass=pkicertificatetemplate)
+        (!(mspki-enrollment-flag:1.2.840.113556.1.4.804:=2))
+        (|(mspki-ra-signature=0)(!(mspki-ra-signature=*)))
```

Comment:
I don't *think* these fields are necessary. If a signature or manager approval is required but we have write access to the template, we can change it to remove the requirement. I think we just need the `objectclass` filter so we know what objects we're looking at.